Mathlib/FieldTheory/SeparableClosure.lean

```diff
@@ -0,0 +1,386 @@
+    [h1 : IsSeparable F L1] [h2 : IsSeparable F L2] :
+    IsSeparable F (L1 ⊔ L2 : IntermediateField F E) := by
+  rw [← le_separableClosure_iff] at h1 h2 ⊢
+  exact sup_le h1 h2
+
+/-- A compositum of separable extensions is separable. -/
+instance IntermediateField.isSeparable_iSup {ι : Type*} {t : ι → IntermediateField F E}
+    [h : ∀ i, IsSeparable F (t i)] : IsSeparable F (⨆ i, t i : IntermediateField F E) := by
+  simp_rw [← le_separableClosure_iff] at h ⊢
+  exact iSup_le h
+
+end separableClosure
+
+namespace Field
+
+/-- The (infinite) separable degree for a general field extension `E / F` is defined
+to be the degree of `(separableClosure F E) / F`. -/
+def sepDegree := Module.rank F (separableClosure F E)
+
+/-- The (infinite) inseparable degree for a general field extension `E / F` is defined
+to be the degree of `E / (separableClosure F E)`. -/
+def insepDegree := Module.rank (separableClosure F E) E
+
+/-- The (finite) inseparable degree for a general field extension `E / F` is defined
+to be the degree of `E / (separableClosure F E)` as a natural number. It is defined to be zero
+if such field extension is infinite. -/
+def finInsepDegree : ℕ := Cardinal.toNat (insepDegree F E)
+
+instance instNeZeroSepDegree : NeZero (sepDegree F E) := ⟨rank_pos.ne'⟩
+
+instance instNeZeroInsepDegree : NeZero (insepDegree F E) := ⟨rank_pos.ne'⟩
+
+instance instNeZeroFinInsepDegree [FiniteDimensional F E] :
+    NeZero (finInsepDegree F E) := ⟨finrank_pos.ne'⟩
+
+/-- If `E` and `K` are isomorphic as `F`-algebras, then they have the same (infinite)
+separable degree over `F`. -/
+theorem sepDegree_eq_of_equiv (i : E ≃ₐ[F] K) :
+    Cardinal.lift.{w} (sepDegree F E) = Cardinal.lift.{v} (sepDegree F K) :=
+  (separableClosure.algEquivOfAlgEquiv F E K i).toLinearEquiv.lift_rank_eq
+
+/-- The (infinite) separable degree multiply by the (infinite) inseparable degree is equal
+to the (infinite) field extension degree. -/
+theorem sepDegree_mul_insepDegree : sepDegree F E * insepDegree F E = Module.rank F E :=
+  rank_mul_rank F (separableClosure F E) E
+
+/-- If `E` and `K` are isomorphic as `F`-algebras, then they have the same (infinite)
+inseparable degree over `F`. -/
+theorem insepDegree_eq_of_equiv (i : E ≃ₐ[F] K) :
+    Cardinal.lift.{w} (insepDegree F E) = Cardinal.lift.{v} (insepDegree F K) :=
+  Algebra.lift_rank_eq_of_equiv_equiv (separableClosure.algEquivOfAlgEquiv F E K i).toRingEquiv
+    i.toRingEquiv (by rfl)
+
+/-- If `E` and `K` are isomorphic as `F`-algebras, then they have the same (finite)
+inseparable degree over `F`. -/
+theorem finInsepDegree_eq_of_equiv (i : E ≃ₐ[F] K) :
+    finInsepDegree F E = finInsepDegree F K := by
+  simpa only [Cardinal.toNat_lift] using congr_arg Cardinal.toNat
+    (insepDegree_eq_of_equiv F E K i)
+
+@[simp]
+theorem sepDegree_self : sepDegree F F = 1 := by
+  have : separableClosure F F = ⊥ := toSubalgebra_injective <| Subsingleton.elim _ _
+  rw [sepDegree, this, IntermediateField.rank_bot]
```

Comment:
Maybe we should transport the Subalgebra Subsingleton instance to IntermediateField, so you can just `rw [Subsingleton.elim (separableClosure F F) ⊥`.